### DISCUSSION TITLE
Handle camera limits inside a scene

### DISCUSSION THREAD

**Discussion by zen14774** [ramokz/phantom-camera#157] [Question / Help Wanted] [answered]
I create a scene for a room that contains a PhantomCamera2D and I set the camera limits so it is confined to the walls of the room. Then I add this scene to my main scene and when the player enters the room I give that camera max priority.

This shows the camera in the wrong position, I think it's because the limits applied to the Camera2D are local to the scene that contains the PhantomCamera2D and not converted to global coordinates. What's the correct way to handle this?

**Comment by ramokz** [comment]
Just tried it out by adding an external scene with a `PCam2D` to a parent scene.
By the looks of it, it does seem to act as if they're all in the same scene without any strange positional issues with the camera.

<img width="1658" alt="image" src="https://github.com/ramokz/phantom-camera/assets/5159399/b8ebaf23-04a7-443d-8cde-547739fe02b5">

> `Main Scene`. Existing `PCam2D` has a priorrity of 5.

<br>

<img width="1673" alt="image" src="https://github.com/ramokz/phantom-camera/assets/5159399/dbc795f9-22f8-4370-8c19-3b507e60e517">

> External scene added to `Main Scene`. This `PCam2D` has a priority of 30, and controls the `Camera2D` node when added to the `Main Scene`. Notice the Limit size in the `Main Scene` reflecting the properties in this scene.

**Reply by ramokz** [reply]
It sounds like it might be related to the position of the `PCam2D` within the child scene.

Is this happening both in the editor _and_ while running the scene, or just one of them?
And would you also be able to share either a small sample project with the issue in question, or screenshots of your scene structure? So to have a better idea about how it's been put together.

**Comment by zen14774** [comment]
For context, I'm trying to get the camera to work similar to the first 10 seconds of this video: https://www.youtube.com/watch?v=aUQWezmUS_g&t=2s

Here is a small scene showing what I have gotten to work so far: [cameratest.zip](https://github.com/ramokz/phantom-camera/files/13767047/cameratest.zip)

I managed to get a bit closer by converting the limits to global coordinates, that's the code `room.gd` ready method. Without that conversion the limits used are global and the camera points at the wrong position in the main scene. However, doing it that way means the camera doesn't tween when moving between "rooms", like it does in the video.

**Reply by ramokz** [reply]
Ah I can see what you're trying to do now.

Yes, the Limit values are global coordinates, which means you would have to manually calculate the position of the individual room's position within the main scene's global space, then grab its size and then dynamically set the `Left`, `Top`, `Right` and `Bottom` properties based on those two values. A bit cumbersome, but durable.

_However_, if you're using a `TileMap`, at least based on the sample project you shared, you can alternatively use the `TileMap Limit` property, which would do it automatically. The only requirement for it to work is to make sure the `TileMap` is big enough for the limit to exceed the camera's viewport. At the minute the `TileMap` is just the floor, so width-wise it's fine, but it also needs some additional height to allow the `Camera2D` to actually move around.

Have attached an updated version of the project you shared that uses the `TileMap Limit` approach with some additional bits - mainly by adding transitional `PCam2Ds` on either side of the room. It's fairly rough, and needs a lot more polish, but _should_ be close to what you wanted I think:
[cameratest_2.zip](https://github.com/ramokz/phantom-camera/files/13767582/cameratest_2.zip)

**Reply by zen14774** [reply]
That's better, but the transition still snaps sometimes. In top-down view the rooms have four sides and different sizes and the player can move diagonally, so I would like to keep the scene setup simple.

I am using TileMaps in most rooms so that's awesome. For a couple that I don't use TileMaps I can set the limits manually as I did before, but it'd be great if there was a way to constrain the camera to an area, similar to what we can do with paths but for a Shape2D?

I've tried tweening the camera limits. I modified this code of `PhantomCamera2D`:
```gdscript
func _set_camera_2d_limit(side: int, limit: int) -> void:
	_has_valid_pcam_owner()
	if not is_active(): return
	var cam = get_pcam_host_owner().camera_2D
	var tween = get_tree().create_tween()
	tween.tween_method(func(x): cam.set_limit(side, x), cam.get_limit(side), limit, get_tween_duration())
```
This works pretty well, the only issue I see is that the camera has weird acceleration for a moment when the two tweens overlap.

[cameratest_3.zip](https://github.com/ramokz/phantom-camera/files/13771444/cameratest_3.zip)

**Reply by zen14774** [reply]
@ramokz  I tried the new version, thanks for adding the new limit node!

Unfortunately, I can't seem to get it to work the way I needed with the transitions. For reference, this is what the previous version looks like with my modification:

https://github.com/ramokz/phantom-camera/assets/43204260/925731ca-5ff2-4853-80cc-26d80622614f

This is what it's doing now:

https://github.com/ramokz/phantom-camera/assets/43204260/99d853b2-9bdc-4c5e-a677-30756f6e5ce8

My modified code doesn't seem to work with the new version, so I'll update when I have more time to look into it. I've disabled the update pop up for now.

**Reply by ramokz** [reply]
Have just published a new release, 0.6.1, which adds support for `Shape2D` resources inside `CollisionShape2D` nodes to control the `Limit` size, thought it was a good idea! It should now also disable the `Limit` size while tweens are happening to avoid unintended behaviour.

I would be cautious and even advise against tweening the `Camera2D` `Limit`. Depending on how the different limits overlap, the limit change can push the `Camera2D` if the limit sides catch up to and moves past it while the `Camera2D` tween is happening. It ultimately makes the tweening result fairly unpredictable. That is _likely_ what is causing of the acceleration you see.

Wonder if the simplest solution would be to allow the `PCam2D` to optionally be clamped to the `Limit` area when it is inactive or tweening. Upon entering a room, it should then assign the player as its `Follow Target` and it will automatically be positioned and confined to the room's `Limit` area, like what the `Camera2D` ultimately will be, all while the tweening is happening. Afterwards, the `Camera2D`'s `Limit` system will be reapplied and the `PCam2D` gets its `Limit` clamp disabled as it doesn't need it any more. That _should_ mean you can use a single `PCam2D` per room without applying any additional logic beyond triggering the priority changes.

I can share an updated version of the `PhantomCamera2D.gd` here for you to try out before merging anything into the main branch, if that sound sensible?

**Reply by zen14774** [reply]
You're right that tweening the limits pushes the camera around and looks wonky, as seen in the first video.

If I'm understanding correctly the solution you propose starts the tween with the camera in the current position and tweens towards where the camera will end when the new limits are applied? That should avoid the 'jump' in the second example (which is using 0.6.1) where during the tween the camera jumps towards the character?

If you share the code I'm happy to try it out in my project and update the minimal example.

**Reply by ramokz** [reply]
Yeah, so in the current version the `PCam2D` itself doesn't care about the `Limit` system, and so is never constrained by it and merely just passes the data on to the `Camera2D` as it becomes active. The solve here is to essentially apply the `Limit` values to the `PCam2D` while it's inactive, so it matches how the `Camera2D` will eventually act like once the tween is complete.

Have added an updated version of the scripts below. Since a few files were changed, you can just swap the entire `addon/phantom_camera/scripts` directory with the one attached below, and it should just work.
[phantom_camera:scripts.zip](https://github.com/ramokz/phantom-camera/files/13798333/phantom_camera.scripts.zip)

Have also published a 0.6.2 branch, which has an updated example scene that showcases the tweening behaviour.

**Reply by ramokz** [reply]
Another update to try, it reverts the smoothing system back to the original approach, i.e. pre-0.6, due to the `Position Smoothing` from 0.6 caused a persistent conflict with how the tweening system works. This version _should_ eliminate all jitter from transitions and thereafter, but let me know if you still see any!
[phantom_camera:scripts.zip](https://github.com/ramokz/phantom-camera/files/13802217/phantom_camera.scripts.zip)

**Reply by zen14774** [reply]
I tried this last one and it works very well:

https://github.com/ramokz/phantom-camera/assets/43204260/417fbf0e-99a7-45e2-8897-db23a137c880

**Reply by zen14774** [reply]
Sorry for nitpicking like this, but I noticed something and I don't have to look more into it today, so I wanted to mention it:

When setting the tween duration to be short (0.25) it seems to overshoot the tween for a couple of frames. I captured a new video and split it into frames. You can see that the movement is linear but on frame 78 it goes over a bit, then on 79 it bounces back into the final position and stays there.

[cameratest_4.zip](https://github.com/ramokz/phantom-camera/files/13814273/cameratest_4.zip)

75: ![frames075](https://github.com/ramokz/phantom-camera/assets/43204260/bf1dcf14-d032-426b-bee3-d88735da26a7)
76: ![frames076](https://github.com/ramokz/phantom-camera/assets/43204260/af9b96ef-fc34-478d-afc6-16601851fa46)
77:
![frames077](https://github.com/ramokz/phantom-camera/assets/43204260/ed804a29-9aa1-4a84-8538-47d0a08042a8)
78:
![frames078](https://github.com/ramokz/phantom-camera/assets/43204260/39f6b657-e0df-473a-8269-9a861fd1b482)
79:
![frames079](https://github.com/ramokz/phantom-camera/assets/43204260/cd4f5c78-a88b-4f2c-a80d-d34d9843e451)
80:
![frames080](https://github.com/ramokz/phantom-camera/assets/43204260/8809b9ad-617f-40e9-9982-f62eac2685f6)

**Reply by ramokz** [reply, marked as answer]
Nitpicks are great! Thanks for pointing this out, would probably never have spotted this myself.

Attached a copy of the project, with a slightly tweaked `phantom_camera_host.gd`, which should remove the additional tween frame.

[cameratest_5.zip](https://github.com/ramokz/phantom-camera/files/13814403/cameratest_5.zip)

**Reply by zen14774** [reply]
This version works like a charm. Many thanks!

**Reply by ramokz** [reply]
Thanks for finding and help fixing those issues!
